Add combined fee and price route

vkgnosis opened the pull request:
Part of https://github.com/gnosis/gp-v2-services/issues/679 . See that issue for motivation.

In this PR I add a naive implementation of the combined route. This is what afaik the front end does at the moment just done by the backend. After this PR we can improve the implementation to take the amounts into account like with Felix' more accurate fee route.

I would like to discuss how the route should look like before I add the openapi documentation. I have chosen the structure of the fee route which passes all the parameters through the query. This is different from the markets route which passes them in the path. I prefer the query version.

I haven't added the price impact that is mentioned in the issue because I wasn't sure how to correctly calculate it.

I chose a new route instead of a v2 of an existing route because I the new name is clearer but I don't mind changing this.

### Test Plan
new unit tests and manual test (TODO)

vkgnosis committed:
Add combined fee and price route

fleupold commented on orderbook/src/api/get_fee_and_price.rs:
```diff
@@ -0,0 +1,186 @@
+    expiration_date: DateTime<Utc>,
+    min_fee: U256,
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    amount_after_fee: BigInt,
```
This is the same as on the markets route, but isn't it a bit weird to have min_fee and amount_after_fee be different types?

fleupold commented on orderbook/src/api/get_fee_and_price.rs:
```diff
@@ -0,0 +1,186 @@
+}
+
+fn request() -> impl Filter<Extract = (Query,), Error = Rejection> + Clone {
+    warp::path!("fee_and_price")
```
nit: you are not really getting a price, but an amount (proceeds or costs) - we could also consider calling it "quote" in case maybe other relevant data may be added later (but I also like the very clear name you chose)?

fleupold commented on orderbook/src/api/get_fee_and_price.rs:
```diff
@@ -0,0 +1,186 @@
+    query: Query,
+) -> Result<Response, Error> {
+    let (min_fee, expiration_date) = fee_calculator
+        .min_fee(query.sell_token.0, None, None, None)
```
This should not use the legacy API but instead set buyToken, amount and kind (since it already knows it).

vkgnosis replied:
This is intentional to make the implementation simpler. With the advanced implementation we need to take into account that the fee for the full sell amount is different from the new fee after deducting the full amount fee. So I want to handle that in a future PR.

fleupold replied:
The frontend currently uses the new API (with the full sell amount which is technically not correct but will still give a better estimate than not providing any information about the route at all). It doesn't have to be accurate as in the "verification" we still use the legacy API (effectively accepting the lowest fee that was provided over all routes/amounts for this sell token). Yet it would be nice if it was as close as possible (the same the frontend uses today)

If left like this, it would regress the amount of fee the orders are providing by default.

vkgnosis replied:
I didn't realize the verification uses the legacy api. Fixed.

fleupold commented on orderbook/src/api/get_fee_and_price.rs:
```diff
@@ -0,0 +1,186 @@
+        .await?;
+    let amount_after_fee = match query.kind {
+        OrderKind::Buy => query.amount.to_big_rational() * price,
+        OrderKind::Sell => query.amount.saturating_sub(min_fee).to_big_rational() / price,
```
Maybe to be more accurate we can do the subtraction before we estimate the price (but you also mentioned correctness is not the primary goal of this PR)?

fleupold commented on orderbook/src/api/get_fee_and_price.rs:
```diff
@@ -0,0 +1,186 @@
+    sell_token: H160Wrapper,
+    buy_token: H160Wrapper,
+    #[serde(with = "u256_decimal")]
+    amount: U256,
```
nit: maybe this could highlight that this is the `amount_before_fee`?

fleupold commented on orderbook/src/api/get_fee_and_price.rs:
```diff
@@ -0,0 +1,186 @@
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Response {
```
I liked about the markets route that it tells you which token the amounts refer to (to make confusing the API's response harder). Not sure this is needed for the fee (which is always sell token) but I do think it would be nice for the amount.

vkgnosis replied:
For this route I think it makes more sense to not specify the token as the amount is always expressed in the other token with respect to the input amount. Adding the token only introduces ambiguity.
 I have added comments to the struct explain this.

bh2smith reviewed (approved):
Looks pretty good although I was not really part of the development for the two things being combined here so I'd say that @fleupold's comments seem appropriate. Once all this is complete, would we also be eliminating the possibility to query for these two things individually through the API?

vkgnosis commented:
Once the frontend uses the new route we should definitely remove the existing fee route as it is superseeded by this one. Not sure about the markets route yet.

vkgnosis commented:
Addressed the comments.
Added out_amount_before_fees to the response because it is likely useful for the frontend so they can pick which price they want to display.

anxolin commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+#[derive(Deserialize)]
+struct Query {
+    #[serde(with = "h160_hexadecimal")]
+    sell_token: H160,
```
Why the fee uses `H160Wrapper` and here `H160`? 
Is the wrapper already H160 with serialize/deserialize impl?

vkgnosis replied:
The wrapper only matter for serialization so it's not something you would see from a http standpoint. The difference is that the wrapper implement the rust trait FromStr while this serde_with only works for serialization.

anxolin commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Response {
+    expiration_date: DateTime<Utc>,
```
Would this be cleat that the expiration applies only to the fee? should fee and expiration be in a nested object to make this relation clearer?

vkgnosis replied:
I would definitely make it clear in the swagger docs. I don't mind making a nested struct either.

anxolin replied:
I would favour the nested struct, like `struct Response { fee:  {  expiration_date: DateTime<Utc>, min_fee: U256.. } ... }`

This way, is self explanatory just by the data

vkgnosis replied:
used nested struct

anxolin commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+    // than out_amount_before_fees because the fee is first deducted from the sell amount before
+    // applying the price.
+    #[serde(with = "u256_decimal")]
+    out_amount_after_fees: U256,
```
Nice names. Hopefully this kill any ambiguity

anxolin replied:
EDIT: Now im not sure 😅, will for buy orders be actually really after fees? My understanding is that is not.

vkgnosis replied:
As the comment says, for buy orders the amount before and after fee is the same.
The reason is that for buy orders the fee is in the sell token but we don't know how much the user has in the sell token. We only know they want want to buy x amount in the buy token. So what is returned is the fee they need to set and the sell amount they need set, but there is nothing to deduct. Does that make sense?

anxolin replied:
Yep, it does. It's hard to find infalible names when we play with fees, prices, and buy/sell, but I think is good enough

anxolin commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+
+#[derive(Debug)]
+enum Error {
+    NotFound,
```
I find weird this error. I notice it before, but never mentioned. I think the error should give a better context:
 - NotExistentToken, InssuficientLiquidity, ... (or whatever the error is modelling)

vkgnosis replied:
Agree this should be clearer.

vkgnosis replied:
fixed

anxolin commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+}
+
+fn request() -> impl Filter<Extract = (Query,), Error = Rejection> + Clone {
+    warp::path!("fee_and_price")
```
Is this the path? 

I would literally use this one: 
![image](https://user-images.githubusercontent.com/2352112/122577032-9f185600-d052-11eb-9005-31b28c893dfd.png)
https://protocol-rinkeby.dev.gnosisdev.com/api/#/default/get_api_v1_markets__baseToken___quoteToken___kind___amount_

It just changes the signature. But to make a gradual change, we can just create `/markets/next/...`  and after we migrate the frontends (balancer and cowswap) we kill the other two endpoints and we leave only the markers one. 

Would that make sense?

vkgnosis replied:
I don't have a strong opinion on the path but I like the new one a little more. The bigger change with this route is that it doesn't put the parameters in the path like the markets route. Instead it puts them in the query like the fee route. Do you have a preference there?

anxolin replied:
I found `/markets/<base>-<quote>/sell/<amount>` and `/markets/<base>-<quote>/buy/<amount>` very declarative, but maybe is a personal preference. 

I usually like APIs to be hierarchical, like `/markets/<base>-<quote>` modeling a market, and then with the these endpoints we get a sell/buy quote. The only difference is that now we return the fee together with the quote, which imo makes sense to be part of it. 

I let other give their opinions. I won't be opposed to the new name if that is the general sentiment.

anxolin commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+    buy_token: H160,
+    // For sell orders in sell token, for buy orders in the buy token.
+    #[serde(with = "u256_decimal")]
+    in_amount_before_fees: U256,
```
I like the name. Although for buy orders, the name might be confusing, cause it's actually after fees :)

vkgnosis replied:
I am starting to think we might want to routes to remove all ambiguity. One for selling and one for buying.

anxolin replied:
If it's not too much overhead, it would be ideal yep.

anxolin reviewed (commented):
Good call! yes, we can try to adapt to this once is final, and kill the other endpoints. 

About price impact, it's a bit complex, we can leave it out, or even try to do it in the front end.
My understanding is that we estimate in USD the amounts before fees, and see the the relation between them. 

Something like:

sell orders impact = `(USD_ESTIMATION(in_amount_before_fees * price) - USD_ESTIMATION(in_amount_before_fees)) / USD_ESTIMATION(in_amount_before_fees * price)`


buy orders impact = `(USD_ESTIMATION(in_amount_before_fees) - USD_ESTIMATION(in_amount_before_fees / price) / USD_ESTIMATION(in_amount_before_fees)`


* USD price estimation, we can use some approximation like how much is 1 WETH worth of a token in some stable token. Suggesting 1 WETH cause I belive Martin was convinced of that amount, but probably we can study what's a good value, and if maybe we can set an amount in USD instead of WETH to make it simpler

nlordell reviewed (approved):
Code looks good to me.

vkgnosis committed:
separate buy sell routes

vkgnosis commented:
I changed the api to have two separate routes `fee_and_quote/sell` and  `fee_and_quote/buy` with separate input and output structs. I feel this makes it as clear as possible how this works and resolves confusion around input and output amounts working differently depending on type of order.
After doing this I also decided to get rid of the redundant information buy_amount_before_fee, buy_amount_after_fee. The comments make it clear what the values mean and users of the api can calculate derived statistics themself. I stuck with sell_amount_before_fee and buy_amount_after fee so that the routes are symmetrical.

vkgnosis commented on orderbook/src/api/get_fee_and_quote.rs:
```diff
@@ -0,0 +1,238 @@
+    #[serde(with = "h160_hexadecimal")]
+    buy_token: H160,
+    // For sell orders in sell token, for buy orders in the buy token.
+    #[serde(with = "u256_decimal")]
```
serde_as is more powerful but also harder to understand. I prefer to use the simpler approach when possible and don't want to consistently use serde_as because it will make it less clear when serde_as is needed and switching away from it later potentially harder.

nlordell replied:
I also prefer `serde(with = ...)` (but only because its more familiar to me :stuck_out_tongue:). The comment was more understanding why we need both side by side (which can be a little confusing when looking at the codebase and trying to decide which one to use).

nlordell reviewed (approved):
Splitting them into two routes makes sense and helps avoid confusion.

fleupold reviewed (approved):
Very nice and clear routing! Will you update the openapi in a separate PR?

vkgnosis commented:
I was waiting for consensus on the route. It seems like we're happy with this so I am going to add openapi to this PR before merging.

vkgnosis commented:
Added openapi docs and am testing manually right now. Any last thoughts @anxolin before merging?

anxolin commented on orderbook/openapi.yml:
```diff
@@ -296,6 +296,84 @@ paths:
           description: Token non-existent or no valid price found
         500:
           description: Unexpected internal error while processing the request
+  /api/v1/fee_and_quote/sell:
```
I'm still not excited about the name, although I don't consider super important, and since there's many approvals already I assume people is happy with it, so I won't oppose

I'm approving, although leaving my thinking so it can get lost in this PR comments :) 
- why snake case? I thought we were using kebab 🍗
- I liked more how it was before, base-quote (in terms of a market), although i see for example paraswap don't use the market notation. The reason i like market notation is because for me is closer to what a user does: you are in WETH-DAI market, and you want to buy or sell WETH.
- Even if we don't like the market notation, and we want to use sell/buy token notation because it's closer to the contracts, I would prefer a simpler name `prices/sell`. For me, the fact that we return also the fee shouldn't have so much relevance to change the endpoint name. The name doesn't need to be an exaustive description of all the data. The main thing we are querying is the price.

vkgnosis replied:
1. We are using kebab case for the types but hasn't come up yet for routes because the rest are one word. Actually, there is solvable_orders which is snake case but I don't think we consciously thought about that. I will change this PR to use kebab case for the route too.
2. Personally I find calling things sell_amount and buy_amount more obvious than quote and base. But not sure how normal this is. I'm also not sure the market abstraction is that useful because the price and fee in one direction can be (slightly) different than in the other direction.
3. Right now we have a fee route and a markets route so I wanted this name to show this is route is different. Maybe once it is clear which routes will stay we can rename this one and think about what the best name would be. Like you said I also don't find it super important and don't want to spend too long discussing here.

nlordell replied:
> Personally I find calling things sell_amount and buy_amount more obvious than quote and base

I think this is fine. We used "markets" in GPv1 for price estimation and it was a pretty large source of confusion. I think making it "sell" and "buy" might be weird to the financially inclined, but at least to me it feels less ambiguous and you can immediately populate request data without having to think too much.

> I will change this PR to use kebab case for the route too.

Personally, I prefer kebab case for URLs. I don't know why, nor do I have a good reason for it.

anxolin replied:
All good 👌
Thxxx

anxolin reviewed (approved):
Good to go! Thanks for checking with us @e00E !!
